File: python/seal_gate/engine/input_normalizer.py
from ..types import SealInput, SealEvidence
from ..errors import SealInputError

VALID_ARTIFACT_TYPES = {'llm_response', 'code_diff', 'test_plan', 'design', 'migration'}
VALID_RISK_LEVELS = {'LOW', 'MEDIUM', 'HIGH', 'CRITICAL'}


def normalize(raw: dict) -> SealInput:
    output = raw.get('output', '')
    if not output or not str(output).strip():
        raise SealInputError('output is required and must be non-empty', 'output')

    artifact_type = raw.get('artifact_type', '')
    if artifact_type not in VALID_ARTIFACT_TYPES:
        raise SealInputError(f'artifact_type must be one of: {", ".join(VALID_ARTIFACT_TYPES)}', 'artifact_type')

    risk_hint = raw.get('risk_hint')
    if risk_hint is not None and risk_hint not in VALID_RISK_LEVELS:
        raise SealInputError(f'risk_hint must be one of: {", ".join(VALID_RISK_LEVELS)} or null', 'risk_hint')

    evidence_raw = raw.get('evidence') or {}
    evidence = SealEvidence(
        test_log=evidence_raw.get('test_log', ''),
        build_log=evidence_raw.get('build_log', ''),
        diff=evidence_raw.get('diff', ''),
        references=evidence_raw.get('references', []),
    )

    return SealInput(
        artifact_type=artifact_type,
        spec=raw.get('spec'),
        output=str(output).strip(),
        evidence=evidence,
        risk_hint=risk_hint,
        context=raw.get('context'),
    )

File: src/engine/input-normalizer.ts
import { SealInput, ArtifactType, RiskLevel } from '../types.ts'
import { SealInputError } from '../errors.ts'

const VALID_ARTIFACT_TYPES: ArtifactType[] = ['llm_response', 'code_diff', 'test_plan', 'design', 'migration', 'plan_review']
const VALID_RISK_LEVELS: RiskLevel[] = ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL']

export class InputNormalizer {
  static normalize(input: Partial<SealInput>): SealInput {
    if (!input.output || input.output.trim() === '') {
      throw new SealInputError('output is required and must be non-empty', 'output')
    }
    if (!input.artifact_type || !VALID_ARTIFACT_TYPES.includes(input.artifact_type)) {
      throw new SealInputError(`artifact_type must be one of: ${VALID_ARTIFACT_TYPES.join(', ')}`, 'artifact_type')
    }
    if (input.risk_hint !== null && input.risk_hint !== undefined && !VALID_RISK_LEVELS.includes(input.risk_hint)) {
      throw new SealInputError(`risk_hint must be one of: ${VALID_RISK_LEVELS.join(', ')} or null`, 'risk_hint')
    }
    return {
      artifact_type: input.artifact_type,
      spec: input.spec ?? null,
      output: input.output.trim(),
      evidence: {
        test_log: input.evidence?.test_log ?? '',
        build_log: input.evidence?.build_log ?? '',
        diff: input.evidence?.diff ?? '',
        references: input.evidence?.references ?? [],
      },
      risk_hint: input.risk_hint ?? null,
      context: input.context,
    }
  }
}
